File: code/apps/web/src/lib/billing/mapping.ts
/**
 * Billing protobuf messages to the desk's models.
 *
 * The one thing this layer must not do is turn money into a number. The wire
 * carries int64 minor units, which arrive as a bigint; everything above keeps
 * them as a bigint, and there is no point in this file where an amount becomes
 * a float.
 */
import { timestampDate } from '@bufbuild/protobuf/wkt';
import type { Timestamp } from '@bufbuild/protobuf/wkt';
import {
	ChargeStatus as WireChargeStatus,
	EntryKind as WireEntryKind,
	InvoiceStatus as WireInvoiceStatus,
	PaymentMethod as WirePaymentMethod,
	type Charge as WireCharge,
	type CloseException as WireCloseException,
	type Invoice as WireInvoice,
	type LedgerEntry as WireLedgerEntry,
	type Money as WireMoney
} from '$gen/healthcare/billing/v1/billing_pb.js';
import {
	presentCharge,
	presentEntry,
	presentInvoice,
	type ChargeStatus,
	type CloseException,
	type EntryKind,
	type InvoiceStatus,
	type PaymentMethod,
	type PresentedCharge,
	type PresentedEntry,
	type PresentedInvoice
} from './account.js';
import type { Money } from './money.js';

function toDate(timestamp: Timestamp | undefined): Date | null {
	return timestamp ? timestampDate(timestamp) : null;
}

/**
 * An amount off the wire.
 *
 * A missing Money message becomes zero *in the account's currency*, supplied by
 * the caller — never an empty currency string, which would then refuse to sum
 * with everything else and turn a blank field into a screen-wide failure.
 */
export function toMoney(amount: WireMoney | undefined, fallbackCurrency: string): Money {
	if (!amount) {
		return { minor: 0n, currency: fallbackCurrency.toUpperCase() };
	}
	return {
		minor: amount.minor,
		currency: (amount.currency || fallbackCurrency).toUpperCase()
	};
}

/** The wire shape for an amount, for building a request. */
export function fromMoney(amount: Money): { minor: bigint; currency: string } {
	return { minor: amount.minor, currency: amount.currency };
}

const entryKinds: Record<WireEntryKind, EntryKind> = {
	[WireEntryKind.UNSPECIFIED]: 'unspecified',
	[WireEntryKind.INVOICE]: 'invoice',
	[WireEntryKind.PAYMENT]: 'payment',
	[WireEntryKind.REFUND]: 'refund',
	[WireEntryKind.DEPOSIT]: 'deposit',
	[WireEntryKind.DEPOSIT_APPLIED]: 'deposit_applied',
	[WireEntryKind.WRITE_OFF]: 'write_off',
	[WireEntryKind.ADJUSTMENT]: 'adjustment'
};

const invoiceStatuses: Record<WireInvoiceStatus, InvoiceStatus> = {
	[WireInvoiceStatus.UNSPECIFIED]: 'unspecified',
	[WireInvoiceStatus.DRAFT]: 'draft',
	[WireInvoiceStatus.ISSUED]: 'issued',
	[WireInvoiceStatus.SUPERSEDED]: 'superseded',
	[WireInvoiceStatus.CANCELLED]: 'cancelled'
};

const chargeStatuses: Record<WireChargeStatus, ChargeStatus> = {
	[WireChargeStatus.UNSPECIFIED]: 'unspecified',
	[WireChargeStatus.POSTED]: 'posted',
	[WireChargeStatus.INVOICED]: 'invoiced',
	[WireChargeStatus.VOIDED]: 'voided',
	[WireChargeStatus.HELD]: 'held'
};

const methods: Record<WirePaymentMethod, PaymentMethod> = {
	[WirePaymentMethod.UNSPECIFIED]: 'unspecified',
	[WirePaymentMethod.CASH]: 'cash',
	[WirePaymentMethod.CARD]: 'card',
	[WirePaymentMethod.UPI]: 'upi',
	[WirePaymentMethod.BANK_TRANSFER]: 'bank_transfer',
	[WirePaymentMethod.CHEQUE]: 'cheque',
	[WirePaymentMethod.PAYER_SETTLEMENT]: 'payer_settlement'
};

/** The wire enum for a payment method, for building a request. */
export const wireMethods: Record<PaymentMethod, WirePaymentMethod> = {
	unspecified: WirePaymentMethod.UNSPECIFIED,
	cash: WirePaymentMethod.CASH,
	card: WirePaymentMethod.CARD,
	upi: WirePaymentMethod.UPI,
	bank_transfer: WirePaymentMethod.BANK_TRANSFER,
	cheque: WirePaymentMethod.CHEQUE,
	payer_settlement: WirePaymentMethod.PAYER_SETTLEMENT
};

/** Adapts a wire LedgerEntry. */
export function toPresentedEntry(entry: WireLedgerEntry, currency: string): PresentedEntry {
	return presentEntry({
		entryId: entry.entryId,
		kind: entryKinds[entry.kind] ?? 'unspecified',
		amount: toMoney(entry.amount, currency),
		receiptNumber: entry.receiptNumber,
		method: methods[entry.method] ?? 'unspecified',
		reason: entry.reason,
		occurredAt: toDate(entry.occurredAt) ?? new Date(0),
		recordedBy: entry.recordedBy
	});
}

/** Adapts a wire Invoice. */
export function toPresentedInvoice(invoice: WireInvoice, currency: string): PresentedInvoice {
	return presentInvoice({
		invoiceId: invoice.invoiceId,
		number: invoice.number || invoice.invoiceId.slice(0, 8),
		status: invoiceStatuses[invoice.status] ?? 'unspecified',
		total: toMoney(invoice.total, currency),
		issuedAt: toDate(invoice.issuedAt),
		supersededBy: invoice.supersededBy,
		correctsInvoiceId: invoice.correctsInvoiceId,
		lineCount: invoice.lines.length
	});
}

/** Adapts a wire Charge. */
export function toPresentedCharge(charge: WireCharge, currency: string): PresentedCharge {
	return presentCharge({
		chargeId: charge.chargeId,
		description: charge.description || charge.serviceCode || 'Charge',
		department: charge.department,
		quantity: charge.quantity,
		total: toMoney(charge.total, currency),
		status: chargeStatuses[charge.status] ?? 'unspecified',
		covered: charge.covered,
		coverageNote: charge.coverageNote,
		occurredAt: toDate(charge.occurredAt) ?? toDate(charge.postedAt) ?? new Date(0)
	});
}

/** Adapts a wire CloseException. */
export function toCloseException(
	exception: WireCloseException,
	currency: string
): CloseException {
	return {
		check: exception.check,
		detail: exception.detail,
		// Only when the server actually sent one. A zero rendered beside "two
		// services are not yet invoiced" reads as "nothing outstanding", which
		// is the opposite of what the exception says.
		amount: exception.amount ? toMoney(exception.amount, currency) : null,
		references: exception.references
	};
}

File: code/apps/mobile/lib/src/billing/mapping.dart
/// Wire types to the billing screens' models (UX-W1-06).
///
/// Same unknown-enum discipline as the other three workspaces; the reasoning is
/// in `reception/mapping.dart`.
///
/// One extra rule here, and it is the one this whole workspace exists to keep:
/// **money crosses this boundary as an integer and never as anything else.**
/// The proto carries `int64` minor units, `fixnum.Int64` arrives in Dart, and
/// `.toInt()` is exact on the 64-bit platforms this app ships to. Nothing in
/// this file divides, rounds, or touches a `double`.
library;

import 'package:protobuf/protobuf.dart' as pb;

import '../gen/healthcare/billing/v1/billing.pb.dart' as wire;
import 'account.dart';
import 'money.dart';

/// Field tags carrying the enums this module reads. Pinned by test.
const int entryKindField = 5; // billing.v1.LedgerEntry.kind
const int entryMethodField = 10; // billing.v1.LedgerEntry.method
const int invoiceStatusField = 4; // billing.v1.Invoice.status
const int chargeStatusField = 24; // billing.v1.Charge.status

/// True when [message] carried a value for [tag] this build could not read.
bool sentUnknownValueFor(pb.GeneratedMessage message, int tag) =>
    message.unknownFields.hasField(tag);

/// An amount, wire to model.
///
/// `toInt()` rather than any arithmetic: the minor units are already the exact
/// integer the ledger holds, and every operation that could lose a digit is one
/// this module declines to perform.
Money moneyOf(wire.Money amount, {String fallbackCurrency = ''}) => Money(
      minor: amount.minor.toInt(),
      // A currency-less amount is possible on the wire (a zero, usually) and a
      // blank currency renders as a bare number, which on a receipt is worse
      // than a wrong one — it cannot even be questioned.
      currency: amount.currency.isNotEmpty
          ? amount.currency.toUpperCase()
          : fallbackCurrency.toUpperCase(),
    );

EntryKind entryKindOf(wire.LedgerEntry entry) {
  if (sentUnknownValueFor(entry, entryKindField)) {
    return EntryKind.unrecognised;
  }
  return switch (entry.kind) {
    wire.EntryKind.ENTRY_KIND_INVOICE => EntryKind.invoice,
    wire.EntryKind.ENTRY_KIND_PAYMENT => EntryKind.payment,
    wire.EntryKind.ENTRY_KIND_REFUND => EntryKind.refund,
    wire.EntryKind.ENTRY_KIND_DEPOSIT => EntryKind.deposit,
    wire.EntryKind.ENTRY_KIND_DEPOSIT_APPLIED => EntryKind.depositApplied,
    wire.EntryKind.ENTRY_KIND_WRITE_OFF => EntryKind.writeOff,
    wire.EntryKind.ENTRY_KIND_ADJUSTMENT => EntryKind.adjustment,
    wire.EntryKind.ENTRY_KIND_UNSPECIFIED => EntryKind.unspecified,
    _ => EntryKind.unrecognised,
  };
}

PaymentMethod methodOf(wire.LedgerEntry entry) {
  if (sentUnknownValueFor(entry, entryMethodField)) {
    return PaymentMethod.unrecognised;
  }
  return switch (entry.method) {
    wire.PaymentMethod.PAYMENT_METHOD_CASH => PaymentMethod.cash,
    wire.PaymentMethod.PAYMENT_METHOD_CARD => PaymentMethod.card,
    wire.PaymentMethod.PAYMENT_METHOD_BANK_TRANSFER => PaymentMethod.bankTransfer,
    wire.PaymentMethod.PAYMENT_METHOD_UPI => PaymentMethod.upi,
    wire.PaymentMethod.PAYMENT_METHOD_CHEQUE => PaymentMethod.cheque,
    wire.PaymentMethod.PAYMENT_METHOD_PAYER_SETTLEMENT =>
      PaymentMethod.payerSettlement,
    wire.PaymentMethod.PAYMENT_METHOD_UNSPECIFIED => PaymentMethod.unspecified,
    _ => PaymentMethod.unrecognised,
  };
}

/// Wire method for a model one, for taking a payment.
///
/// Returns null for anything that is not a real method, which
/// `validatePayment` has already refused — the second reading is here because
/// this is where it would otherwise reach the wire.
wire.PaymentMethod? wireMethodOf(PaymentMethod method) => switch (method) {
      PaymentMethod.cash => wire.PaymentMethod.PAYMENT_METHOD_CASH,
      PaymentMethod.card => wire.PaymentMethod.PAYMENT_METHOD_CARD,
      PaymentMethod.bankTransfer =>
        wire.PaymentMethod.PAYMENT_METHOD_BANK_TRANSFER,
      PaymentMethod.upi => wire.PaymentMethod.PAYMENT_METHOD_UPI,
      PaymentMethod.cheque => wire.PaymentMethod.PAYMENT_METHOD_CHEQUE,
      PaymentMethod.payerSettlement =>
        wire.PaymentMethod.PAYMENT_METHOD_PAYER_SETTLEMENT,
      PaymentMethod.unspecified || PaymentMethod.unrecognised => null,
    };

InvoiceStatus invoiceStatusOf(wire.Invoice invoice) {
  if (sentUnknownValueFor(invoice, invoiceStatusField)) {
    // Not "draft". A status this build cannot read must never make an invoice
    // editable — the patient may be holding a copy of it.
    return InvoiceStatus.unrecognised;
  }
  return switch (invoice.status) {
    wire.InvoiceStatus.INVOICE_STATUS_DRAFT => InvoiceStatus.draft,
    wire.InvoiceStatus.INVOICE_STATUS_ISSUED => InvoiceStatus.issued,
    wire.InvoiceStatus.INVOICE_STATUS_SUPERSEDED => InvoiceStatus.superseded,
    wire.InvoiceStatus.INVOICE_STATUS_CANCELLED => InvoiceStatus.cancelled,
    wire.InvoiceStatus.INVOICE_STATUS_UNSPECIFIED => InvoiceStatus.unspecified,
    _ => InvoiceStatus.unrecognised,
  };
}

ChargeStatus chargeStatusOf(wire.Charge charge) {
  if (sentUnknownValueFor(charge, chargeStatusField)) {
    // Not "posted". Counting an unreadable charge as unbilled would add it to
    // a total somebody bills from.
    return ChargeStatus.unrecognised;
  }
  return switch (charge.status) {
    wire.ChargeStatus.CHARGE_STATUS_POSTED => ChargeStatus.posted,
    wire.ChargeStatus.CHARGE_STATUS_INVOICED => ChargeStatus.invoiced,
    wire.ChargeStatus.CHARGE_STATUS_VOIDED => ChargeStatus.voided,
    wire.ChargeStatus.CHARGE_STATUS_HELD => ChargeStatus.held,
    wire.ChargeStatus.CHARGE_STATUS_UNSPECIFIED => ChargeStatus.unspecified,
    _ => ChargeStatus.unrecognised,
  };
}

PresentedEntry entryOf(wire.LedgerEntry entry, {String currency = ''}) =>
    presentEntry(
      entryId: entry.entryId,
      kind: entryKindOf(entry),
      amount: moneyOf(entry.amount, fallbackCurrency: currency),
      occurredAt: entry.occurredAt.toDateTime().toUtc(),
      receiptNumber: entry.receiptNumber,
      method: methodOf(entry),
      reason: entry.reason,
      recordedBy: entry.recordedBy,
    );

PresentedInvoice invoiceOf(wire.Invoice invoice, {String currency = ''}) =>
    presentInvoice(
      invoiceId: invoice.invoiceId,
      number: invoice.number,
      status: invoiceStatusOf(invoice),
      total: moneyOf(invoice.total, fallbackCurrency: currency),
      // hasIssuedAt rather than a zero check: a draft has no issue date, and
      // the epoch shown as one reads as an invoice issued in 1970.
      issuedAt:
          invoice.hasIssuedAt() ? invoice.issuedAt.toDateTime().toUtc() : null,
      supersededBy: invoice.supersededBy,
      correctsInvoiceId: invoice.correctsInvoiceId,
      lineCount: invoice.lines.length,
    );

PresentedCharge chargeOf(wire.Charge charge, {String currency = ''}) =>
    presentCharge(
      chargeId: charge.chargeId,
      display: charge.description.isNotEmpty
          ? charge.description
          : charge.serviceCode,
      status: chargeStatusOf(charge),
      total: moneyOf(charge.total, fallbackCurrency: currency),
      occurredAt: charge.hasOccurredAt()
          ? charge.occurredAt.toDateTime().toUtc()
          : null,
    );

CloseException closeExceptionOf(wire.CloseException exception,
        {String currency = ''}) =>
    CloseException(
      check: exception.check_1,
      detail: exception.detail,
      amount: exception.hasAmount()
          ? moneyOf(exception.amount, fallbackCurrency: currency)
          : null,
      references: List.of(exception.references),
    );
